File: lib/Services/ReportService.dart
// lib/Services/ReportService.dart
import 'dart:io';
import 'package:http/http.dart' as http;
import 'package:path_provider/path_provider.dart';
import 'package:open_filex/open_filex.dart';
import 'package:flutter/foundation.dart' show kIsWeb;
import 'dart:html' as html;
import 'api_constants.dart';
import 'Authservices.dart';

class ReportService {
  static final ReportService _instance = ReportService._internal();
  factory ReportService() => _instance;
  ReportService._internal();

  final AuthService _authService = AuthService.instance;

  /// Download Patient Report PDF
  Future<Map<String, dynamic>> downloadPatientReport(String patientId) async {
    try {
      final token = await _authService.getToken();
      if (token == null) {
        return {
          'success': false,
          'message': 'Authentication token not found. Please login again.'
        };
      }

      final url = '${ApiConstants.baseUrl}/api/reports-proper/patient/$patientId';
      
      final response = await http.get(
        Uri.parse(url),
        headers: {
          'Authorization': 'Bearer $token',
          'Content-Type': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        // Get filename from header or create default
        String filename = 'Patient_Report_${DateTime.now().millisecondsSinceEpoch}.pdf';
        
        final contentDisposition = response.headers['content-disposition'];
        if (contentDisposition != null) {
          final filenameMatch = RegExp(r'filename="?([^"]+)"?').firstMatch(contentDisposition);
          if (filenameMatch != null) {
            filename = filenameMatch.group(1) ?? filename;
          }
        }

        if (kIsWeb) {
          // Web platform - trigger download
          final blob = html.Blob([response.bodyBytes]);
          final url = html.Url.createObjectUrlFromBlob(blob);
          final anchor = html.AnchorElement(href: url)
            ..setAttribute('download', filename)
            ..click();
          html.Url.revokeObjectUrl(url);

          return {
            'success': true,
            'message': 'Patient report downloaded successfully',
            'filename': filename
          };
        } else {
          // Mobile/Desktop platform - save to file
          final directory = await getApplicationDocumentsDirectory();
          final file = File('${directory.path}/$filename');
          await file.writeAsBytes(response.bodyBytes);

          // Try to open the file
          try {
            await OpenFilex.open(file.path);
          } catch (e) {
            print('Could not open file: $e');
          }

          return {
            'success': true,
            'message': 'Patient report saved successfully',
            'path': file.path,
            'filename': filename
          };
        }
      } else if (response.statusCode == 404) {
        return {
          'success': false,
          'message': 'Patient not found'
        };
      } else {
        return {
          'success': false,
          'message': 'Failed to generate patient report: ${response.statusCode}'
        };
      }
    } catch (e) {
      print('Error downloading patient report: $e');
      return {
        'success': false,
        'message': 'Error downloading patient report: $e'
      };
    }
  }

  /// Download Doctor Report PDF
  Future<Map<String, dynamic>> downloadDoctorReport(String doctorId) async {
    try {
      final token = await _authService.getToken();
      if (token == null) {
        return {
          'success': false,
          'message': 'Authentication token not found. Please login again.'
        };
      }

      final url = '${ApiConstants.baseUrl}/api/reports-proper/doctor/$doctorId';
      
      final response = await http.get(
        Uri.parse(url),
        headers: {
          'Authorization': 'Bearer $token',
          'Content-Type': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        // Get filename from header or create default
        String filename = 'Doctor_Report_${DateTime.now().millisecondsSinceEpoch}.pdf';
        
        final contentDisposition = response.headers['content-disposition'];
        if (contentDisposition != null) {
          final filenameMatch = RegExp(r'filename="?([^"]+)"?').firstMatch(contentDisposition);
          if (filenameMatch != null) {
            filename = filenameMatch.group(1) ?? filename;
          }
        }

        if (kIsWeb) {
          // Web platform - trigger download
          final blob = html.Blob([response.bodyBytes]);
          final url = html.Url.createObjectUrlFromBlob(blob);
          final anchor = html.AnchorElement(href: url)
            ..setAttribute('download', filename)
            ..click();
          html.Url.revokeObjectUrl(url);

          return {
            'success': true,
            'message': 'Doctor report downloaded successfully',
            'filename': filename
          };
        } else {
          // Mobile/Desktop platform - save to file
          final directory = await getApplicationDocumentsDirectory();
          final file = File('${directory.path}/$filename');
          await file.writeAsBytes(response.bodyBytes);

          // Try to open the file
          try {
            await OpenFilex.open(file.path);
          } catch (e) {
            print('Could not open file: $e');
          }

          return {
            'success': true,
            'message': 'Doctor report saved successfully',
            'path': file.path,
            'filename': filename
          };
        }
      } else if (response.statusCode == 404) {
        return {
          'success': false,
          'message': 'Doctor not found'
        };
      } else {
        return {
          'success': false,
          'message': 'Failed to generate doctor report: ${response.statusCode}'
        };
      }
    } catch (e) {
      print('Error downloading doctor report: $e');
      return {
        'success': false,
        'message': 'Error downloading doctor report: $e'
      };
    }
  }

  /// Download Staff Report PDF
  Future<Map<String, dynamic>> downloadStaffReport(String staffId) async {
    try {
      final token = await _authService.getToken();
      if (token == null) {
        return {
          'success': false,
          'message': 'Authentication token not found. Please login again.'
        };
      }

      final url = '${ApiConstants.baseUrl}/api/reports-proper/staff/$staffId';
      
      final response = await http.get(
        Uri.parse(url),
        headers: {
          'Authorization': 'Bearer $token',
          'Content-Type': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        // Get filename from header or create default
        String filename = 'Staff_Report_${DateTime.now().millisecondsSinceEpoch}.pdf';
        
        final contentDisposition = response.headers['content-disposition'];
        if (contentDisposition != null) {
          final filenameMatch = RegExp(r'filename="?([^"]+)"?').firstMatch(contentDisposition);
          if (filenameMatch != null) {
            filename = filenameMatch.group(1) ?? filename;
          }
        }

        if (kIsWeb) {
          // Web platform - trigger download
          final blob = html.Blob([response.bodyBytes]);
          final url = html.Url.createObjectUrlFromBlob(blob);
          final anchor = html.AnchorElement(href: url)
            ..setAttribute('download', filename)
            ..click();
          html.Url.revokeObjectUrl(url);

          return {
            'success': true,
            'message': 'Staff report downloaded successfully',
            'filename': filename
          };
        } else {
          // Mobile/Desktop platform - save to file
          final directory = await getApplicationDocumentsDirectory();
          final file = File('${directory.path}/$filename');
          await file.writeAsBytes(response.bodyBytes);

          // Try to open the file
          try {
            await OpenFilex.open(file.path);
          } catch (e) {
            print('Could not open file: $e');
          }

          return {
            'success': true,
            'message': 'Staff report saved successfully',
            'path': file.path,
            'filename': filename
          };
        }
      } else if (response.statusCode == 404) {
        return {
          'success': false,
          'message': 'Staff member not found'
        };
      } else {
        return {
          'success': false,
          'message': 'Failed to generate staff report: ${response.statusCode}'
        };
      }
    } catch (e) {
      print('Error downloading staff report: $e');
      return {
        'success': false,
        'message': 'Error downloading staff report: $e'
      };
    }
  }
}

File: react/hms/src/services/reportService.js
/**
 * reportService.js
 * React equivalent of Flutter's ReportService.dart
 * Handles PDF report generation and download for various entities
 */

import axios from 'axios';
import { ReportsEndpoints } from './apiConstants';

/**
 * Get auth token from localStorage
 */
const getAuthToken = () => {
  return localStorage.getItem('auth_token') || localStorage.getItem('x-auth-token') || localStorage.getItem('authToken');
};

/**
 * Create axios instance with auth headers
 */
const createAxiosInstance = () => {
  const token = getAuthToken();
  return axios.create({
    baseURL: process.env.REACT_APP_API_URL || 'https://hms-dev.onrender.com/api',
    headers: {
      'Content-Type': 'application/json',
      ...(token && {
        'x-auth-token': token,
        'Authorization': `Bearer ${token}`
      })
    }
  });
};

class ReportService {
  /**
   * Download Patient Report PDF
   * @param {string} patientId - The patient ID
   * @returns {Promise<Object>} Response with success status and message
   */
  async downloadPatientReport(patientId) {
    try {
      const axiosInstance = createAxiosInstance();
      const response = await axiosInstance.get(
        ReportsEndpoints.patientReport(patientId),
        {
          responseType: 'blob', // Important for PDF download
        }
      );

      // Extract filename from content-disposition header
      const contentDisposition = response.headers['content-disposition'];
      let filename = `Patient_Report_${Date.now()}.pdf`;

      if (contentDisposition) {
        const filenameMatch = contentDisposition.match(/filename="?([^";]+)"?/);
        if (filenameMatch && filenameMatch[1]) {
          filename = filenameMatch[1];
        }
      }

      // Create blob and trigger download
      const blob = new Blob([response.data], { type: 'application/pdf' });
      const url = window.URL.createObjectURL(blob);
      const link = document.createElement('a');
      link.href = url;
      link.setAttribute('download', filename);
      document.body.appendChild(link);
      link.click();
      link.remove();
      window.URL.revokeObjectURL(url);

      return {
        success: true,
        message: 'Patient report downloaded successfully',
        filename,
      };
    } catch (error) {
      console.error('Error downloading patient report:', error);
      return {
        success: false,
        message: error.response?.data?.message || error.message || 'Failed to download patient report',
      };
    }
  }

  /**
   * Download Doctor Report PDF
   * @param {string} doctorId - The doctor ID
   * @returns {Promise<Object>} Response with success status and message
   */
  async downloadDoctorReport(doctorId) {
    try {
      const axiosInstance = createAxiosInstance();
      const response = await axiosInstance.get(
        ReportsEndpoints.doctorReport(doctorId),
        {
          responseType: 'blob',
        }
      );

      const contentDisposition = response.headers['content-disposition'];
      let filename = `Doctor_Report_${Date.now()}.pdf`;

      if (contentDisposition) {
        const filenameMatch = contentDisposition.match(/filename="?([^";]+)"?/);
        if (filenameMatch && filenameMatch[1]) {
          filename = filenameMatch[1];
        }
      }

      const blob = new Blob([response.data], { type: 'application/pdf' });
      const url = window.URL.createObjectURL(blob);
      const link = document.createElement('a');
      link.href = url;
      link.setAttribute('download', filename);
      document.body.appendChild(link);
      link.click();
      link.remove();
      window.URL.revokeObjectURL(url);

      return {
        success: true,
        message: 'Doctor report downloaded successfully',
        filename,
      };
    } catch (error) {
      console.error('Error downloading doctor report:', error);
      return {
        success: false,
        message: error.response?.data?.message || error.message || 'Failed to download doctor report',
      };
    }
  }

  /**
   * Download Staff Report PDF
   * @param {string} staffId - The staff ID
   * @returns {Promise<Object>} Response with success status and message
   */
  async downloadStaffReport(staffId) {
    try {
      const axiosInstance = createAxiosInstance();
      const response = await axiosInstance.get(
        ReportsEndpoints.staffReport(staffId),
        {
          responseType: 'blob',
        }
      );

      const contentDisposition = response.headers['content-disposition'];
      let filename = `Staff_Report_${Date.now()}.pdf`;

      if (contentDisposition) {
        const filenameMatch = contentDisposition.match(/filename="?([^";]+)"?/);
        if (filenameMatch && filenameMatch[1]) {
          filename = filenameMatch[1];
        }
      }

      const blob = new Blob([response.data], { type: 'application/pdf' });
      const url = window.URL.createObjectURL(blob);
      const link = document.createElement('a');
      link.href = url;
      link.setAttribute('download', filename);
      document.body.appendChild(link);
      link.click();
      link.remove();
      window.URL.revokeObjectURL(url);

      return {
        success: true,
        message: 'Staff report downloaded successfully',
        filename,
      };
    } catch (error) {
      console.error('Error downloading staff report:', error);
      return {
        success: false,
        message: error.response?.data?.message || error.message || 'Failed to download staff report',
      };
    }
  }

  /**
   * Download Prescription PDF
   * @param {string} prescriptionId - The prescription ID
   * @param {string} patientName - The patient name (for filename)
   * @returns {Promise<Object>} Response with success status and message
   */
  async downloadPrescription(prescriptionId, patientName) {
    try {
      const axiosInstance = createAxiosInstance();
      const response = await axiosInstance.get(
        ReportsEndpoints.prescriptionPdf(prescriptionId),
        {
          responseType: 'blob',
        }
      );

      const contentDisposition = response.headers['content-disposition'];
      let filename = `Prescription_${patientName}_${Date.now()}.pdf`;

      if (contentDisposition) {
        const filenameMatch = contentDisposition.match(/filename="?([^";]+)"?/);
        if (filenameMatch && filenameMatch[1]) {
          filename = filenameMatch[1];
        }
      }

      const blob = new Blob([response.data], { type: 'application/pdf' });
      const url = window.URL.createObjectURL(blob);
      const link = document.createElement('a');
      link.href = url;
      link.setAttribute('download', filename);
      document.body.appendChild(link);
      link.click();
      link.remove();
      window.URL.revokeObjectURL(url);

      return {
        success: true,
        message: 'Prescription downloaded successfully',
        filename,
      };
    } catch (error) {
      console.error('Error downloading prescription:', error);
      return {
        success: false,
        message: error.response?.data?.message || error.message || 'Failed to download prescription',
      };
    }
  }

  /**
   * Get pathology reports list
   * @returns {Promise<Object>} Response with reports data
   */
  async getPathologyReports() {
    try {
      const axiosInstance = createAxiosInstance();
      const response = await axiosInstance.get(ReportsEndpoints.pathologyReports);
      return {
        success: true,
        reports: response.data?.reports || response.data || [],
      };
    } catch (error) {
      console.error('Error fetching pathology reports:', error);
      return {
        success: false,
        message: error.response?.data?.message || error.message || 'Failed to fetch reports',
        reports: [],
      };
    }
  }

  /**
   * View PDF or Image in a new tab
   * Uses the public PDF endpoint for easy browser viewing
   * @param {string} fileId - The file ID from database (pdfRef or imageRef)
   */
  viewPdf(fileId) {
    if (!fileId) {
      console.error('No file ID provided');
      alert('No file available to view');
      return;
    }
    const baseUrl = process.env.REACT_APP_API_URL || 'https://hms-dev.onrender.com/api';
    const url = `${baseUrl}/scanner-enterprise/pdf-public/${fileId}`;
    console.log('📄 Opening file in new tab:', url);
    window.open(url, '_blank');
  }

  /**
   * View image in a modal or new tab
   * @param {string} imageId - The image ID from database
   */
  viewImage(imageId) {
    if (!imageId) {
      console.error('No image ID provided');
      alert('No image available to view');
      return;
    }
    const baseUrl = process.env.REACT_APP_API_URL || 'https://hms-dev.onrender.com/api';
    const url = `${baseUrl}/scanner-enterprise/pdf-public/${imageId}`;
    console.log('🖼️ Opening image in new tab:', url);
    window.open(url, '_blank');
  }

  /**
   * Download test report file
   * @param {FormData} formData - Form data with file and metadata
   * @returns {Promise<Object>} Response with success status
   */
  async uploadTestReport(formData) {
    try {
      const token = getAuthToken();
      const axiosInstance = axios.create({
        headers: {
          'Content-Type': 'multipart/form-data',
          ...(token && { 'Authorization': `Bearer ${token}` })
        }
      });

      const response = await axiosInstance.post(
        ReportsEndpoints.uploadTestReport,
        formData
      );

      return {
        success: true,
        message: 'Test report uploaded successfully',
        data: response.data,
      };
    } catch (error) {
      console.error('Error uploading test report:', error);
      return {
        success: false,
        message: error.response?.data?.message || error.message || 'Failed to upload test report',
      };
    }
  }
}

// Export singleton instance
const reportService = new ReportService();
export default reportService;
